src/utils-dispatch.c

```diff
@@ -0,0 +1,105 @@
+#include "vctrs.h"
+#include "utils.h"
+
+
+// Defined below
+enum vctrs_class_type class_type(SEXP x);
+static enum vctrs_class_type class_type_impl(SEXP class);
+static const char* class_type_as_str(enum vctrs_class_type type);
+
+
+// [[ register() ]]
+SEXP vctrs_class_type(SEXP x) {
+  return Rf_mkString(class_type_as_str(class_type(x)));
+}
+
+// [[ include("utils.h") ]]
+bool is_data_frame(SEXP x) {
+  enum vctrs_class_type type = class_type(x);
+  return
+    type == vctrs_class_bare_data_frame ||
+    type == vctrs_class_bare_tibble ||
+    type == vctrs_class_data_frame;
+}
+
+// [[ include("utils.h") ]]
+bool is_native_df(SEXP x) {
+  enum vctrs_class_type type = class_type(x);
+  return
+    type == vctrs_class_bare_data_frame ||
+    type == vctrs_class_bare_tibble;
+}
+
+// [[ include("utils.h") ]]
+bool is_bare_data_frame(SEXP x) {
+  return class_type(x) == vctrs_class_bare_data_frame;
+}
+
+// [[ include("utils.h") ]]
+bool is_bare_tibble(SEXP x) {
+  return class_type(x) == vctrs_class_bare_tibble;
+}
+
+// [[ include("utils.h") ]]
+bool is_record(SEXP x) {
+  enum vctrs_class_type type = class_type(x);
+  return
+    type == vctrs_class_rcrd ||
+    type == vctrs_class_posixlt;
+}
+
+
+enum vctrs_class_type class_type(SEXP x) {
+  if (!OBJECT(x)) {
+    return vctrs_class_none;
+  }
+
+  SEXP class = PROTECT(Rf_getAttrib(x, R_ClassSymbol));
+  enum vctrs_class_type type = class_type_impl(class);
+
+  UNPROTECT(1);
+  return type;
+}
+
+static enum vctrs_class_type class_type_impl(SEXP class) {
+  int n = Rf_length(class);
+  SEXP const* class_ptr = STRING_PTR(class);
+  SEXP const* p = class_ptr;
+
+  switch (n) {
```

Comment:
I think this needs a comment like "First check for data frames"

Reply:
This checks for bare classes, i.e. those for which we know exactly how many strings compose their class. Good point that this needs comments.